osu.Server.Spectator/Hubs/ScoreUploader.cs

```diff
@@ -0,0 +1,163 @@
+            timerCancellationToken = timerCancellationSource.Token;
+
+            timer = new Timer(5000);
+            timer.AutoReset = false;
+            timer.Elapsed += (_, _) => Task.Run(Flush);
+            timer.Start();
+        }
+
+        /// <summary>
+        /// Enqueues a new score to be uploaded.
+        /// </summary>
+        /// <param name="token">The score's token.</param>
+        /// <param name="score">The score.</param>
+        public void Enqueue(long token, Score score)
+        {
+            if (!shouldSaveReplays)
+                return;
+
+            Interlocked.Increment(ref remainingUsages);
+
+            var cancellation = new CancellationTokenSource();
+            cancellation.CancelAfter(TimeSpan.FromMilliseconds(TimeoutInterval));
+
+            queue.Enqueue(new UploadItem(token, score, cancellation));
+        }
+
+        /// <summary>
+        /// Flushes all pending uploads.
+        /// </summary>
+        public async Task Flush()
+        {
+            try
+            {
+                timer.Stop();
+
+                if (queue.IsEmpty)
+                    return;
+
+                Console.WriteLine($"Beginning upload of {queue.Count} scores");
+
+                using (var db = databaseFactory.GetInstance())
+                {
+                    int countToTry = queue.Count;
+
+                    for (int i = 0; i < countToTry; i++)
+                    {
+                        if (!queue.TryDequeue(out var item))
+                            continue;
+
+                        long? scoreId = await db.GetScoreIdFromToken(item.Token);
+
+                        if (scoreId == null && !item.Cancellation.IsCancellationRequested)
+                        {
+                            // Score is not ready yet - enqueue for the next attempt.
+                            queue.Enqueue(item);
+                            continue;
+                        }
+
+                        try
+                        {
+                            if (scoreId != null)
+                            {
+                                item.Score.ScoreInfo.OnlineID = scoreId.Value;
+                                await scoreStorage.WriteAsync(item.Score);
+                            }
+                            else
+                                Console.WriteLine($"Score upload timed out for token: {item.Token}");
+                        }
+                        finally
+                        {
+                            item.Dispose();
+                            Interlocked.Decrement(ref remainingUsages);
+                        }
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                Console.WriteLine($"Error during score upload: {e}");
+            }
+            finally
+            {
+                if (timerCancellationToken.IsCancellationRequested)
```

Comment:
Doesn't this mean the server could shut down with items still in the queue?

Reply:
That's what `RemainingUsages` is for. It behaves like every other entity store in waiting for the count to reach 0 before shutdown.